File: leetcode/easy/1.two-sum/solutions/java/solution1.java
import java.util.*;

/**
 * Solution 1: One-Pass Hash Table (Optimal)
 * 
 * Time Complexity: O(n) - single pass through array
 * Space Complexity: O(n) - hash map stores at most n elements
 * 
 * Approach:
 * - Use HashMap to store seen elements (value → index)
 * - For each element, calculate complement = target - current
 * - Check if complement exists in map
 * - If found, return indices; otherwise, add current to map
 */
class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (map.containsKey(complement)) {
                return new int[]{map.get(complement), i};
            }
            map.put(nums[i], i);
        }
        
        return new int[]{}; // Should never reach here
    }
}

File: leetcode/easy/1.two-sum/solutions/go/solution1.go
package main

/**
 * Solution 1: One-Pass Hash Table (Optimal)
 *
 * Time Complexity: O(n) - single pass through array
 * Space Complexity: O(n) - map stores at most n elements
 *
 * Approach:
 * - Use map[int]int to store seen elements (value → index)
 * - For each element, calculate complement = target - current
 * - Check if complement exists in map
 * - If found, return indices; otherwise, add current to map
 *
 * Optimizations:
 * - Pre-allocate map with estimated capacity
 * - Use indexed loop instead of range for better performance
 */
func twoSum(nums []int, target int) []int {
	// Pré-aloca o map com capacidade estimada
	m := make(map[int]int, len(nums))

	for i := 0; i < len(nums); i++ {
		complement := target - nums[i]
		if idx, exists := m[complement]; exists {
			return []int{idx, i}
		}
		m[nums[i]] = i
	}

	return []int{} // Should never reach here
}
